### DISCUSSION TITLE
Adding ability to mosaic nearby tiles when composite_function is NULL in `get_stac_data`

### DISCUSSION THREAD

**Discussion by laurenkwick** [Permian-Global-Research/rsi#76] [Q&A] [answered]
Hi! Starting off by saying thank you for all the work in the package and always helping me out when I have questions :) 

I was wondering if it would make sense to have a way to mosaic neighboring tiles even when `composite_function = NULL` in the `get_stac_data` function. 

This came up for me when I was working with a discrete dataset (ALOS Forest/Non-Forest). My region of interest appears to be larger than the tiles in this dataset. Currently, if a compositing function is not provided, neighboring tiles acquired on the same date will not be mosaic-ed together (which makes sense to me since we would need to define a function for selecting which pixel to use in the case of overlapping pixels!). For discrete datasets which have a temporal dimension (like ALOS FNF which has data from several years) I do not want to apply a compositing function if I am querying across several years. However, I would like the tiles to be mosaic-ed together if my region of interest is large. I have a method for dealing with this at the moment where I sequence through the years available in that dataset and check that the user-entered dates match one of the years the dataset is available before running `get_stac_data`: 

```r
check_date_in_range <- function(start_dt, end_dt, target_yr) {
  # Convert the input strings to Date objects
  start_date <- as.Date(start_dt)
  end_date <- as.Date(end_dt)
  
  # Create the start and end dates for the target year
  target_start <- as.Date(paste0(target_yr,"-01-01"))
  target_end <- as.Date(paste0(target_yr, "-12-31"))
  
  # Check if the date ranges overlap and return the target search range
  if (!(end_date < target_start || start_date > target_end)) {
    return(c(target_start, target_end))
  }
  
  # Otherwise, return NULL
  return(NULL)
}

alos_coll <- "alos-fnf-mosaic"
alos_assets <- "C"
pc_stac_source <- "https://planetarycomputer.microsoft.com/api/stac/v1/"

for (i in seq(2015, 2020, 1)) {
    # Check if query dates contain the ALOS FNF temporal extent
    search_dates <- check_date_in_range(dates[1], dates[2], as.character(i))
    if (is.null(search_dates)) {
      next
    }
    
    tryCatch({
      # Retrieve raster data
      # Add a small buffer to aoi layer to get all pixels within the aoi boundary after reprojection (see rsi::get_stac_data help text)
      image <- c(image, rsi::get_stac_data(
        aoi = sf::st_buffer(aoi_search, dist=50),
        start_date = as.character(search_dates[1]),
        end_date = as.character(search_dates[2]),
        pixel_x_size = 25,
        pixel_y_size = 25,
        asset_names = alos_assets,
        stac_source = pc_stac_source,
        collection = alos_coll,
        sign_function = rsi::sign_planetary_computer,
        rescale_bands=FALSE,
        download_function = rsi::rsi_download_rasters,
        composite_function = "max",
        gdalwarp_options = c("-r", "near", "-multi", "-overwrite", "-co",
                             "COMPRESS=DEFLATE", "-co", "PREDICTOR=2", "-co", "NUM_THREADS=ALL_CPUS"),
      ))
    }, error = function(e) {
      message("No ALOS FNF items were found for ", as.character(i))
    })
  }
  
  if (is.null(image)) {
    stop("No ALOS FNF items were found.")
 }
```

This is working just fine for me! However, I was thinking about this for continuous data like Sentinel-2 for example. Would it be possible to have an argument that supports mosaicing in the spatial dimension even when we don't want to apply a temporal composite? I could see this being difficult if neighboring images weren't acquired on the same date (i.e. 11:59 PM on 01/01/23 and 12:01 AM on 01/02/23 for example). But I don't know how frequently something like that happens. 

I'm interested to see if you have any thoughts on this!

**Comment by mikemahoney218** [comment, marked as answer]
The way you've got things set up has been my general recommendation in the past: determine the time unit that you want to merge items within, create a sequence of those time units spanning your temporal AOI, and make one call to `rsi::get_stac_data` per time unit. 

The reason this isn't inside the package itself is because I think this becomes a rabbit hole really quickly. People often want timeframes like "monthly" or "weekly" which are poorly-defined (is a month 30 days or a single calendar month? Do weeks start Monday or Sunday? and so on and so forth) and are better handled by other packages (I love [lubridate](https://lubridate.tidyverse.org/)), so I'd rather people deal with their temporal chunking using one of those libraries and then bring the output to rsi.

That said, you aren't the first person to ask for this general thing ( https://github.com/Permian-Global-Research/rsi/issues/59, [How Can I](https://permian-global-research.github.io/rsi/articles/How-can-I-.html#get-one-composite-per-year-month-or-other-interval) ), so it's on my mind as a potential future feature... but for now, I'm thinking it's best to outsource the date math.

**Comment by laurenkwick** [comment]
Thank you! I appreciate that response - this will definitely work for my use case
